File: CeliTax/ReceiptBuilder.m
//
//  ReceiptBuilder.m
//  CeliTax
//
//  Created by Leon Chen on 2015-07-20.
//  Copyright (c) 2015 CraveNSave. All rights reserved.
//

#import "ReceiptBuilder.h"

#define kKeyIdentifer       @"identifier"
#define kKeyFileNames       @"filenames"
#define kKeyDateCreated     @"date_created"
#define kKeyTaxYear         @"tax_year"

@implementation ReceiptBuilder

- (Receipt *) buildReceiptFrom: (NSDictionary *) json
{
    if (![json isKindOfClass: [NSDictionary class]])
    {
        return nil;
    }
    
    Receipt *receipt = [[Receipt alloc] init];
    
    receipt.localID = json[kKeyIdentifer];
    receipt.taxYear = [json[kKeyTaxYear] integerValue];
    
    //Filenames
    NSString *filenamesString = json[kKeyFileNames];
    
    NSArray *filenames = [filenamesString componentsSeparatedByString:@","];
    
    receipt.fileNames = [[NSMutableArray alloc] initWithArray: filenames copyItems: NO];
    
    //Data Created
    NSString *dateString = json[kKeyDateCreated];
    
    if (dateString.length)
    {
        NSDateFormatter *gmtDateFormatter = [[NSDateFormatter alloc] init];
        gmtDateFormatter.timeZone = [NSTimeZone timeZoneForSecondsFromGMT:0];
        gmtDateFormatter.dateFormat = @"yyyy-MM-dd HH:mm:ss";
        
        NSDate *dateCreated = [gmtDateFormatter dateFromString: dateString];
        
        receipt.dateCreated = dateCreated;
    }
    
    return receipt;
}

@end

File: CeliTax/ReceiptBuilder.swift
//
//  ReceiptBuilder.swift
//  CeliTax
//
//  Created by Leon Chen on 2015-11-01.
//  Copyright © 2015 CraveNSave. All rights reserved.
//

import Foundation

@objc
class ReceiptBuilder : NSObject //TODO: Remove Subclass to NSObject when the entire app has been converted to Swift
{
    let kKeyIdentifer : String = "identifier"
    let kKeyFileNames : String  = "filenames"
    let kKeyDateCreated : String = "date_created"
    let kKeyTaxYear : String = "tax_year"
    
    func buildReceiptFrom(json : NSDictionary) -> Receipt?
    {
        if (!json.isKindOfClass(NSDictionary))
        {
            return nil
        }
        
        let receipt : Receipt = Receipt()
        
        receipt.localID = json[kKeyIdentifer] as! String
        receipt.taxYear = json[kKeyTaxYear] as! Int
        
        //Filenames
        let filenamesString : NSString = json[kKeyFileNames] as! NSString
        
        let filenames : NSArray = filenamesString.componentsSeparatedByString(",")
        
        receipt.fileNames = NSMutableArray(array: filenames)
        
        //Data Created
        let dateString : NSString = json[kKeyDateCreated] as! NSString
        
        if (dateString.length > 0)
        {
            let gmtDateFormatter : NSDateFormatter = NSDateFormatter()
            gmtDateFormatter.timeZone = NSTimeZone.init(forSecondsFromGMT: 0)
            gmtDateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
            
            let dateCreated : NSDate! = gmtDateFormatter.dateFromString(dateString as String)
            
            receipt.dateCreated = dateCreated
        }
        
        return receipt
    }
}